### PR TITLE
Proto enhancements

### LINKED ISSUE #6224: yul proto spec: add for and switch...case statements
## Abstract

The proto fuzzer does not thoroughly test optimizations on for loop and switch shift statements. This issue tracks support for these statements. Once finished, this should improve proto fuzzers' test coverage

## Motivation

- Extend current proto specification for the said statements
- for loops will be "hardcoded" for a fixed number of iterations e.g.,
```
for {let i := 0x00} lt(i,0x100) { i := add(i, 0x20)}
{
<blockStmt>
}
```
- for loops may contain break/continue statements inside the body of the for loop (see #6136 )

Here's the `for` grammar (taken from the docs): `ForLoop = 'for' Block Expression Block Block`

- switch case is relatively straightforward

## Specification

The technical spec. mirrors [yul grammar](https://solidity.readthedocs.io/en/v0.5.5/yul.html)

### PR DESCRIPTION
**Pull request by bshastry**
### Description

(closes #6224 )

This is a WIP PR that adds the following to yul proto spec
  - for statement
  - switch...case statement
  - shift left/right binary operation

### Checklist
- [x] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

**Comment by chriseth** [comment]
Can you squash this and split into separate pull requests: one per feature and perhaps one that contains all bug fixes?

**Comment by bshastry** [comment]
Okay. I guess one PR per feature would mean one PR each for supporting, (1) For, (2) Switch Case, (3) shift left/right, and (4) string literal+hex literal support?

**Comment by chriseth** [comment]
If this is too fine-grained you can also combine it, but please do not combine features and bugfixes.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 1]
```diff
@@ -23,9 +23,58 @@
 using namespace std;
 using namespace yul::test::yul_fuzzer;
 
+const std::string yul::test::yul_fuzzer::createHex(std::string const& _hexBytes)
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:26]:
```suggestion
std::string const yul::test::yul_fuzzer::createHex(std::string const& _hexBytes)
```
Not completely sure though (about which style is correct)

File: test/tools/ossfuzz/protoToYul.cpp [hunk 2]
```diff
@@ -23,9 +23,58 @@
...
+{
+	std::string tmp = std::string(_hexBytes);
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:28]:
```suggestion
	string tmp = string(_hexBytes);
```

File: test/tools/ossfuzz/protoToYul.cpp [hunk 3]
```diff
@@ -23,9 +23,58 @@
...
+	if (!tmp.empty())
+	{
+		tmp.erase(std::remove_if(tmp.begin(), tmp.end(),
+		                         [](char c) { return !std::isxdigit(c); }), tmp.end());
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:32]:
```suggestion
			[](char c) { return !std::isxdigit(c); }), tmp.end());
```

Comment (leonardoalt) [reply test/tools/ossfuzz/protoToYul.cpp:32]: Also this line is using spaces instead of tabs.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 4]
```diff
@@ -23,9 +23,58 @@
...
+		tmp = tmp.substr(0, 64);
+	}
+	// We need this awkward if case because tmp can also be null post erase call above
+	if (tmp.empty())
+	{
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:37]: Can remove curly brackets

File: test/tools/ossfuzz/protoToYul.cpp [hunk 5]
```diff
@@ -23,9 +23,58 @@
...
+		tmp = "1";
+	}
+	return tmp;
+}
+
+const std::string yul::test::yul_fuzzer::createAlphaNum(std::string const& _strBytes)
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:43]: Similar comments apply to this function.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 6]
```diff
@@ -165,10 +228,10 @@ ostream& yul::test::yul_fuzzer::operator<<(ostream& _os, AssignmentStatement con
 ostream& yul::test::yul_fuzzer::operator<<(ostream& _os, IfStmt const& _x)
 {
 	return _os <<
-	           "if " <<
-	           _x.cond() <<
-	           " " <<
-	           _x.if_body();
+		    "if " <<
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:231]: These lines are using spaces.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 7]
```diff
@@ -210,6 +311,10 @@ ostream& yul::test::yul_fuzzer::operator<<(ostream& _os, Block const& _x)
 		_os << "}\n";
 
 	}
+	else
+	{
```
Comment (leonardoalt) [inline test/tools/ossfuzz/protoToYul.cpp:315]: Can be removed

**Review by leonardoalt — COMMENTED** [review]
I'm not sure how the general style guidelines apply to tools, but in case they do, here are some comments.

File: test/tools/ossfuzz/protoToYul.cpp (hunk 1 above)
Comment (erak) [reply test/tools/ossfuzz/protoToYul.cpp:26]: @leonardoalt I think your suggestion is correct.

File: test/tools/ossfuzz/protoToYul.cpp (hunk 2 above)
Comment (erak) [reply test/tools/ossfuzz/protoToYul.cpp:28]:
```suggestion
	string tmp{_hexBytes};
```

**Comment by bshastry** [comment]
The bug fix portion of this PR has been taken into #6291 

Since that has already been merged, I will rebase and update this PR taking all comments here into account.

File: test/tools/ossfuzz/protoToYul.cpp (hunk 2 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp:28]: `std::string` is a funny story.

The fuzzing toolchain depends on `libc++` (as against libstdc++) and for some reason (although we are doing `using namespace std;`), the linker (clang++) links `string` to a libc++ implementation of it which is slightly different.

This leads to linking errors when we mix (std::string and string implementation in libc++). That's the reason I explicitly reference `std::string`. Other comments are valid though.

File: test/tools/ossfuzz/protoToYul.cpp (hunk 1 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp:26]: Thank you

Comment (chriseth) [reply test/tools/ossfuzz/protoToYul.cpp:26]: I actually don't think returning the string as a const value has any benefit. It actually might add another copy.

**Comment by bshastry** [comment]
Thank you all for your comments. This PR has now been split #6291 #6297 and #6307 . Another separate PR for for loop and switch statement is pending.

Your comments in this PR are not in vain :) I have incorporated them in the individual PRs listed above. I am closing this PR now since it is no longer coherent.

**Pull request closed without merging**